File: ViewModels/Pages/KeyDetectionButtonCardViewModel.vb
Imports System.Runtime.Versioning
Imports System.Windows.Threading
Imports WpfApp1.Infrastructure
Imports WpfApp1.Models
Imports WpfApp1.Services

Namespace ViewModels.Pages
    <SupportedOSPlatform("windows")>
    Public Class KeyDetectionButtonCardViewModel
        Inherits BindableBase
        Implements IDisposable

        Private Const DoublePulseDurationMilliseconds As Double = 350.0

        Private ReadOnly _buttonKind As MouseButtonKind
        Private ReadOnly _stats As InputTimingStatistics
        Private ReadOnly _doublePulseTimer As DispatcherTimer

        Private _title As String
        Private _doubleClickCountText As String
        Private _downCountText As String
        Private _upCountText As String
        Private _currentDownDownText As String
        Private _minimumDownDownText As String
        Private _averageDownDownText As String
        Private _currentDownUpText As String
        Private _minimumDownUpText As String
        Private _averageDownUpText As String
        Private _isPressed As Boolean
        Private _isDoublePulseActive As Boolean
        Private _disposed As Boolean

        Public Sub New(buttonKind As MouseButtonKind)
            _buttonKind = buttonKind
            _stats = New InputTimingStatistics()
            _doublePulseTimer = New DispatcherTimer()
            _doublePulseTimer.Interval = TimeSpan.FromMilliseconds(DoublePulseDurationMilliseconds)
            AddHandler _doublePulseTimer.Tick, AddressOf OnDoublePulseTimerTick

            SyncDisplay()
        End Sub

        Public ReadOnly Property ButtonKind As MouseButtonKind
            Get
                Return _buttonKind
            End Get
        End Property

        Public Property Title As String
            Get
                Return _title
            End Get
            Private Set(value As String)
                SetProperty(_title, value)
            End Set
        End Property

        Public Property DoubleClickCountText As String
            Get
                Return _doubleClickCountText
            End Get
            Private Set(value As String)
                SetProperty(_doubleClickCountText, value)
            End Set
        End Property

        Public Property DownCountText As String
            Get
                Return _downCountText
            End Get
            Private Set(value As String)
                SetProperty(_downCountText, value)
            End Set
        End Property

        Public Property UpCountText As String
            Get
                Return _upCountText
            End Get
            Private Set(value As String)
                SetProperty(_upCountText, value)
            End Set
        End Property

        Public Property CurrentDownDownText As String
            Get
                Return _currentDownDownText
            End Get
            Private Set(value As String)
                SetProperty(_currentDownDownText, value)
            End Set
        End Property

        Public Property MinimumDownDownText As String
            Get
                Return _minimumDownDownText
            End Get
            Private Set(value As String)
                SetProperty(_minimumDownDownText, value)
            End Set
        End Property

        Public Property AverageDownDownText As String
            Get
                Return _averageDownDownText
            End Get
            Private Set(value As String)
                SetProperty(_averageDownDownText, value)
            End Set
        End Property

        Public Property CurrentDownUpText As String
            Get
                Return _currentDownUpText
            End Get
            Private Set(value As String)
                SetProperty(_currentDownUpText, value)
            End Set
        End Property

        Public Property MinimumDownUpText As String
            Get
                Return _minimumDownUpText
            End Get
            Private Set(value As String)
                SetProperty(_minimumDownUpText, value)
            End Set
        End Property

        Public Property AverageDownUpText As String
            Get
                Return _averageDownUpText
            End Get
            Private Set(value As String)
                SetProperty(_averageDownUpText, value)
            End Set
        End Property

        Public Property IsPressed As Boolean
            Get
                Return _isPressed
            End Get
            Private Set(value As Boolean)
                SetProperty(_isPressed, value)
            End Set
        End Property

        Public Property IsDoublePulseActive As Boolean
            Get
                Return _isDoublePulseActive
            End Get
            Private Set(value As Boolean)
                SetProperty(_isDoublePulseActive, value)
            End Set
        End Property

        Public Sub RefreshLocalization(localization As LocalizationManager)
            If localization Is Nothing Then
                Return
            End If

            Title = localization.GetString(ResolveTitleKey())
        End Sub

        Public Sub RegisterDown(timestampMs As Double, doubleClickThresholdMs As Double)
            Dim triggeredDoubleClick = _stats.RegisterDown(timestampMs, doubleClickThresholdMs)
            SyncDisplay()

            If triggeredDoubleClick Then
                PulseDoubleHighlight()
            End If
        End Sub

        Public Sub RegisterUp(timestampMs As Double)
            _stats.RegisterUp(timestampMs)
            SyncDisplay()
        End Sub

        Public Sub ResetStatistics()
            _stats.Reset()
            _doublePulseTimer.Stop()
            IsDoublePulseActive = False
            SyncDisplay()
        End Sub

        Public Sub ResetPressedState()
            _stats.ResetPressedState()
            SyncDisplay()
        End Sub

        Private Sub SyncDisplay()
            DoubleClickCountText = FormatCount(_stats.DoubleClickCount)
            DownCountText = FormatCount(_stats.DownCount)
            UpCountText = FormatCount(_stats.UpCount)
            CurrentDownDownText = FormatMilliseconds(_stats.CurrentDownDownMs)
            MinimumDownDownText = FormatMilliseconds(_stats.MinimumDownDownMs)
            AverageDownDownText = FormatMilliseconds(_stats.AverageDownDownMs)
            CurrentDownUpText = FormatMilliseconds(_stats.CurrentDownUpMs)
            MinimumDownUpText = FormatMilliseconds(_stats.MinimumDownUpMs)
            AverageDownUpText = FormatMilliseconds(_stats.AverageDownUpMs)
            IsPressed = _stats.IsPressed
        End Sub

        Private Function ResolveTitleKey() As String
            Select Case ButtonKind
                Case MouseButtonKind.LeftButton
                    Return "KeyDetection.Mouse.Left"
                Case MouseButtonKind.MiddleButton
                    Return "KeyDetection.Mouse.Middle"
                Case MouseButtonKind.RightButton
                    Return "KeyDetection.Mouse.Right"
                Case MouseButtonKind.ForwardButton
                    Return "KeyDetection.Mouse.Forward"
                Case Else
                    Return "KeyDetection.Mouse.Back"
            End Select
        End Function

        Private Sub PulseDoubleHighlight()
            _doublePulseTimer.Stop()
            IsDoublePulseActive = True
            _doublePulseTimer.Start()
        End Sub

        Private Sub OnDoublePulseTimerTick(sender As Object, e As EventArgs)
            _doublePulseTimer.Stop()
            IsDoublePulseActive = False
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            If _disposed Then
                Return
            End If

            _disposed = True
            _doublePulseTimer.Stop()
            RemoveHandler _doublePulseTimer.Tick, AddressOf OnDoublePulseTimerTick
        End Sub
    End Class
End Namespace

File: src/ClickSyncMouseTester/ViewModels/Pages/KeyDetectionButtonCardViewModel.cs
using ClickSyncMouseTester.Infrastructure;
using ClickSyncMouseTester.Models;
using ClickSyncMouseTester.Services;
using System;
using System.Runtime.Versioning;
using System.Windows.Threading;

namespace ClickSyncMouseTester.ViewModels.Pages;

[SupportedOSPlatform("windows")]
public class KeyDetectionButtonCardViewModel : BindableBase, IDisposable
{
    private const double DoublePulseDurationMilliseconds = 350.0;

    private readonly MouseButtonKind _buttonKind;

    private readonly InputTimingStatistics _stats;

    private readonly DispatcherTimer _doublePulseTimer;

    private string _title;

    private string _doubleClickCountText;

    private string _downCountText;

    private string _upCountText;

    private string _currentDownDownText;

    private string _minimumDownDownText;

    private string _averageDownDownText;

    private string _currentDownUpText;

    private string _minimumDownUpText;

    private string _averageDownUpText;

    private bool _isPressed;

    private bool _isDoublePulseActive;

    private bool _disposed;

    public MouseButtonKind ButtonKind => _buttonKind;

    public string Title
    {
        get
        {
            return _title;
        }
        private set
        {
            SetProperty(ref _title, value, "Title");
        }
    }

    public string DoubleClickCountText
    {
        get
        {
            return _doubleClickCountText;
        }
        private set
        {
            SetProperty(ref _doubleClickCountText, value, "DoubleClickCountText");
        }
    }

    public string DownCountText
    {
        get
        {
            return _downCountText;
        }
        private set
        {
            SetProperty(ref _downCountText, value, "DownCountText");
        }
    }

    public string UpCountText
    {
        get
        {
            return _upCountText;
        }
        private set
        {
            SetProperty(ref _upCountText, value, "UpCountText");
        }
    }

    public string CurrentDownDownText
    {
        get
        {
            return _currentDownDownText;
        }
        private set
        {
            SetProperty(ref _currentDownDownText, value, "CurrentDownDownText");
        }
    }

    public string MinimumDownDownText
    {
        get
        {
            return _minimumDownDownText;
        }
        private set
        {
            SetProperty(ref _minimumDownDownText, value, "MinimumDownDownText");
        }
    }

    public string AverageDownDownText
    {
        get
        {
            return _averageDownDownText;
        }
        private set
        {
            SetProperty(ref _averageDownDownText, value, "AverageDownDownText");
        }
    }

    public string CurrentDownUpText
    {
        get
        {
            return _currentDownUpText;
        }
        private set
        {
            SetProperty(ref _currentDownUpText, value, "CurrentDownUpText");
        }
    }

    public string MinimumDownUpText
    {
        get
        {
            return _minimumDownUpText;
        }
        private set
        {
            SetProperty(ref _minimumDownUpText, value, "MinimumDownUpText");
        }
    }

    public string AverageDownUpText
    {
        get
        {
            return _averageDownUpText;
        }
        private set
        {
            SetProperty(ref _averageDownUpText, value, "AverageDownUpText");
        }
    }

    public bool IsPressed
    {
        get
        {
            return _isPressed;
        }
        private set
        {
            SetProperty(ref _isPressed, value, "IsPressed");
        }
    }

    public bool IsDoublePulseActive
    {
        get
        {
            return _isDoublePulseActive;
        }
        private set
        {
            SetProperty(ref _isDoublePulseActive, value, "IsDoublePulseActive");
        }
    }

    public KeyDetectionButtonCardViewModel(MouseButtonKind buttonKind)
    {
        _buttonKind = buttonKind;
        _stats = new InputTimingStatistics();
        _doublePulseTimer = new DispatcherTimer();
        _doublePulseTimer.Interval = TimeSpan.FromMilliseconds(350.0);
        _doublePulseTimer.Tick += OnDoublePulseTimerTick;
        SyncDisplay();
    }

    public void RefreshLocalization(LocalizationManager localization)
    {
        if (localization != null)
        {
            Title = localization.GetString(ResolveTitleKey());
        }
    }

    public void RegisterDown(double timestampMs, double doubleClickThresholdMs)
    {
        bool isDoubleClick = _stats.RegisterDown(timestampMs, doubleClickThresholdMs);
        SyncDisplay();
        if (isDoubleClick)
        {
            PulseDoubleHighlight();
        }
    }

    public void RegisterUp(double timestampMs)
    {
        _stats.RegisterUp(timestampMs);
        SyncDisplay();
    }

    public void ResetStatistics()
    {
        _stats.Reset();
        _doublePulseTimer.Stop();
        IsDoublePulseActive = false;
        SyncDisplay();
    }

    public void ResetPressedState()
    {
        _stats.ResetPressedState();
        SyncDisplay();
    }

    private void SyncDisplay()
    {
        DoubleClickCountText = KeyDetectionFormatting.FormatCount(_stats.DoubleClickCount);
        DownCountText = KeyDetectionFormatting.FormatCount(_stats.DownCount);
        UpCountText = KeyDetectionFormatting.FormatCount(_stats.UpCount);
        CurrentDownDownText = KeyDetectionFormatting.FormatMilliseconds(_stats.CurrentDownDownMs);
        MinimumDownDownText = KeyDetectionFormatting.FormatMilliseconds(_stats.MinimumDownDownMs);
        AverageDownDownText = KeyDetectionFormatting.FormatMilliseconds(_stats.AverageDownDownMs);
        CurrentDownUpText = KeyDetectionFormatting.FormatMilliseconds(_stats.CurrentDownUpMs);
        MinimumDownUpText = KeyDetectionFormatting.FormatMilliseconds(_stats.MinimumDownUpMs);
        AverageDownUpText = KeyDetectionFormatting.FormatMilliseconds(_stats.AverageDownUpMs);
        IsPressed = _stats.IsPressed;
    }

    private string ResolveTitleKey()
    {
        return ButtonKind switch
        {
            MouseButtonKind.LeftButton => "KeyDetection.Mouse.Left",
            MouseButtonKind.MiddleButton => "KeyDetection.Mouse.Middle",
            MouseButtonKind.RightButton => "KeyDetection.Mouse.Right",
            MouseButtonKind.ForwardButton => "KeyDetection.Mouse.Forward",
            _ => "KeyDetection.Mouse.Back",
        };
    }

    private void PulseDoubleHighlight()
    {
        _doublePulseTimer.Stop();
        IsDoublePulseActive = true;
        _doublePulseTimer.Start();
    }

    private void OnDoublePulseTimerTick(object sender, EventArgs e)
    {
        _doublePulseTimer.Stop();
        IsDoublePulseActive = false;
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _disposed = true;
            _doublePulseTimer.Stop();
            _doublePulseTimer.Tick -= OnDoublePulseTimerTick;
        }
    }

    void IDisposable.Dispose()
    {
        this.Dispose();
    }
}





